import itertools, sys, traceback, pathlib, os, time

###############################
###### Preprocess Config ######
###############################

print("Preprocessing config", flush = True)

# Some standard directories
# If you have a lot of data, you can change this to a larger partition.
# Keep in mind though that the $WRKDIR on turso is very fragile and you will likely shoot down the whole cluster if you do not read and understand this first:
# https://wiki.helsinki.fi/display/it4sci/Lustre+User+Guide#LustreUserGuide-4.1FileLockinginLustreisnotanadvisorytobeignored
DATADIR = "data/"
if "datadir" in config:
    DATADIR = config["datadir"]
DATADIR = os.path.abspath(DATADIR)

# If you have separate executables you can store them here.
PROGRAMDIR = "data/program"
if "programdir" in config:
    PROGRAMDIR = config["programdir"]
PROGRAMDIR = os.path.abspath(PROGRAMDIR)

# You can have a separate dir to store the plots/tables generated by your experiments.
REPORTDIR = "data/reports/"
if "reportdir" in config:
    REPORTDIR = config["reportdir"]
REPORTDIR = os.path.abspath(REPORTDIR)

print("DATADIR: {}".format(DATADIR))
print("PROGRAMDIR: {}".format(PROGRAMDIR))
print("REPORTDIR: {}".format(REPORTDIR))

# This is the maximum number of cores for your default ukko2 machine. Vorna machines have 32.
MAX_THREADS = 56
print("Setting MAX_THREADS to " + str(MAX_THREADS), flush = True)

# If you have source code in your repository, you can make snakemake compile it automatically on change by collecting all the source files.
# This is for Rust code, you can apply it to your programming language of choice.
# rust_sources = list(map(str, itertools.chain(pathlib.Path('implementation').glob('**/Cargo.toml'), pathlib.Path('implementation').glob('**/*.rs'))))

# In case you want to access the current date.
import datetime
today = datetime.date.today().isoformat()

# Chosing the number of CPUs used while running
build_cpus = 1

print("Finished config preprocessing", flush = True)

# Always use conda
workflow.use_conda = True


# import pandas as pd
# # Function to get the collection of fasta files for metagenomes
# def get_metagenome_files(new_folder_name, abundances_file):
#     abundances_df = pd.read_csv(abundances_file, sep='\t')
#     names = abundances_df["Size"]
#     result = []
#     for name in names:
#         if name.endswith(".fna") or name.endswith(".fasta"):
#             new_name = name
#             if new_name.endswith(".fna"): new_name = new_name[:-4] + ".fasta"
#             result.append(os.path.join(DATADIR, new_folder_name, new_name))



############################
###### File Constants ######
############################

# I like to have global constants for most involved file patterns to prevent typos.
# I keep single letter abreviations in front of the wildcards to make the resulting paths more readable.
# Helps especially when there are many arguments.
READS_SRA = os.path.join(DATADIR, "reads", "s{species}", "sra", "reads.sra")
READS_FASTA = os.path.join(DATADIR, "reads", "s{species}", "fasta", "reads.fa")
ASSEMBLY_SOURCE_READS = os.path.join(DATADIR, "assembly", "s{species}-k{k}", "reads.fa")
ASSEMBLY = os.path.join(DATADIR, "assembly", "s{species}-k{k}", "assembly.fa")
REPORT = os.path.join(REPORTDIR, "s{species}-k{k}", "report.txt")
GENOME_ALL_REFERENCES = os.path.join(DATADIR, "{file_name}.fasta")
GENOME_CONCAT_REFERENCES = os.path.join(DATADIR, "{file_name}_concat.fasta")
GENOME_CIRCULAR_REFERENCES = os.path.join(REPORTDIR, "circularization", "{file_name}_k{k}ma{min_abundance}t{threads}", "report.fasta")
# GENOME_CIRCULAR_CONCAT_REFERENCES = os.path.join(REPORTDIR, "circularization", "{file_name}_concat_k{k}ma{min_abundance}t{threads}", "report.fasta")
BUILD_FA = os.path.join(REPORTDIR, "safe_paths_unitigs", "{file_name}_k{k}ma{min_abundance}t{threads}", "report.fasta")
# BUILD_FA = os.path.join(REPORTDIR, "bcalm2", "{file_name}_k{k}ma{min_abundance}t{threads}", "report.fasta")
BUILD_LOG = os.path.join("logs", "build_{file_name}_k{k}ma{min_abundance}t{threads}", "log.log")
NODE_TO_ARC_CENTRIC_DBG_BINARY = os.path.abspath("external_software/node-to-arc-centric-dbg/target/release/node-to-arc-centric-dbg")
NODE_TO_ARC_CENTRIC_DBG = os.path.join(REPORTDIR, "node_to_arc", "{file_name}_k{k}ma{min_abundance}t{threads}", "report.edgelist") #threads 28
SAFE_PATHS_BINARY = os.path.abspath("external_software/safe-paths/target/release/flow_decomposition")
SAFE_PATHS = os.path.join(REPORTDIR, "safe_paths_flowtigs", "{file_name}_k{k}ma{min_abundance}t{threads}", "report.fasta") 
# EXTERNAL_SOFTWARE_ROOTDIR = os.path.join(DATADIR, "external_software")
# QUAST_BINARY = os.path.join(EXTERNAL_SOFTWARE_ROOTDIR, "quast", "quast.py")
QUAST_BINARY = os.path.abspath("external_software/quast/quast.py")
QUAST_OUTPUT_DIR = os.path.join(REPORTDIR, "quast_{algorithm}", "{file_name}_k{k}ma{min_abundance}t{threads}")
QUAST_EXTENDED_OUTPUT_DIR = os.path.join(REPORTDIR, "extended_quast_{algorithm}", "{file_name}_k{k}ma{min_abundance}t{threads}")
PRACTICAL_OMNITIGS_BINARY = os.path.abspath("external_software/practical-omnitigs/implementation/target/release/cli")
PRACTICAL_OMNITIGS = os.path.join(REPORTDIR, "safe_paths_multi-safe", "{file_name}_k{k}ma{min_abundance}t{threads}", "report.fasta")
ECOLI = os.path.join(DATADIR, "ecoli.fasta")
ECOLI_CONCAT = os.path.join(DATADIR, "ecoli_concat.fasta")
SINGLE = os.path.join(DATADIR, "{file_name}.fasta")
SINGLE_CONCAT = os.path.join(DATADIR, "bac_{file_name}_concat.fasta")
# META_BASE7 = os.path.join(DATADIR, "meta_base7.fasta")
# META_BASE7_CONCAT = os.path.join(DATADIR, "meta_base7_concat.fasta")
# MEDIUM20 = os.path.join(DATADIR, "meta_medium20.fasta")
# MEDIUM20_CONCAT = os.path.join(DATADIR, "meta_medium20_concat.fasta")
# COMPLEX32 = os.path.join(DATADIR, "meta_complex32.fasta")
# COMPLEX32_CONCAT = os.path.join(DATADIR, "meta_complex32_concat.fasta")
METAGENOME = os.path.join(DATADIR, "meta_{metagenome}.fasta")
METAGENOME_CONCAT = os.path.join(DATADIR, "meta_{metagenome}_concat.fasta")
PRACTICAL_TEST_OMNITIGS = os.path.join(REPORTDIR, "safe_paths_omnitigs", "{file_name}_k{k}ma{min_abundance}t{threads}", "report.fasta")
PRACTICAL_TRIVIAL_OMNITIGS = os.path.join(REPORTDIR, "safe_paths_trivial-omnitigs", "{file_name}_k{k}ma{min_abundance}t{threads}", "report.fasta")
ALGORITHMS = ["unitigs", "trivial-omnitigs", "multi-safe", "flowtigs", "omnitigs"] # values for the wildcard that chooses which tigs to generate
ALGORITHM_COLUMN_NAMES = ["unitigs", "t. omnitigs", "multi-safe", "flowtigs", "omnitigs"] # column names for the different tigs
FAST_ALGORITHMS = ["unitigs", "trivial-omnitigs", "flowtigs"] # algorithms that have a fast runtime
FAST_ALGORITHM_COLUMN_NAMES = ["unitigs", "t. omnitigs", "flowtigs"] # column names for algorithms that have a fast runtime
CONVERT_VALIDATION_OUTPUTS_TO_LATEX_SCRIPT = "scripts/convert_validation_outputs_to_latex.py"
CONVERT_FAST_VALIDATION_OUTPUTS_TO_LATEX_SCRIPT = "scripts/convert_fast_validation_outputs_to_latex.py"
CREATE_COMBINED_EAXMAX_PLOT_SCRIPT = "scripts/create_combined_eaxmax_plot.py"
REPORT_SUBDIR = os.path.join(REPORTDIR, "final_reports", "final_reports_{file_name}k{k}ma{min_abundance}t{threads}")
REPORT_COMBINED_EAXMAX_PLOT = os.path.join(REPORT_SUBDIR, "combined_eaxmax_plot.pdf")
REPORT_COMBINED_EAXMAX_PLOT_FAST = os.path.join(REPORT_SUBDIR, "combined_eaxmax_plot_fast.pdf")
REPORT_NAME_FILE = os.path.join(REPORT_SUBDIR, "name.txt")
REPORT_HASHDIR = os.path.join(REPORTDIR, "hashdir")
META_BASE7_DIR = os.path.join(DATADIR, "meta", "base7")
# UNPREPROCESSED_META_BASE7 = os.path.join(META_BASE7_DIR, "GCF_0{genome}_genomic.fna")
# UNPREPROCESSED_MEDIUM20 = os.path.join(DATADIR, "meta", "metamedium20", "GCF_{genome}_genomic.fna")
# PREPROCESSED_GENOME = os.path.join(DATADIR, "metabase7", "GCF_0{genome}_genomic.fasta")
# PREPROCESSED_MEDIUM = os.path.join(DATADIR, "metamedium20", "GCF_{genome}_genomic.fasta")
# UNPREPROCESSED_METAGENOME = os.path.join(DATADIR, "meta", "{metagenome}", "{genome}.fna")
UNPREPROCESSED_METAGENOME = os.path.join(DATADIR, "meta", "{metagenome}", "{genome}.fasta")
# PREPROCESSED_METAGENOME = os.path.join(DATADIR, "meta{metagenome}", "{genome}.fasta")
PREPROCESSED_METAGENOME = os.path.join(DATADIR, "preprocessed_metagenome2", "{metagenome}", "{genome}.fasta")
META_BASE7_UNPREPROCESSED = [os.path.join(META_BASE7_DIR, "GCF_000005845.2_ASM584v2_genomic.fna"), os.path.join(META_BASE7_DIR, "GCF_000159115.1_ASM15911v1_genomic.fna"), os.path.join(META_BASE7_DIR, "GCF_000175375.1_ASM17537v1_genomic.fna"), os.path.join(META_BASE7_DIR, "GCF_000376705.1_ASM37670v1_genomic.fna"), os.path.join(META_BASE7_DIR, "GCF_003999335.1_ASM399933v1_genomic.fna"), os.path.join(META_BASE7_DIR, "GCF_017638885.1_ASM1763888v1_genomic.fna"), os.path.join(META_BASE7_DIR, "GCF_019890915.1_ASM1989091v1_genomic.fna")]
# META_BASE7_FASTA = [os.path.join(DATADIR, "metabase7", "GCF_000005845.2_ASM584v2_genomic.fasta"), os.path.join(DATADIR, "metabase7", "GCF_000159115.1_ASM15911v1_genomic.fasta"), os.path.join(DATADIR, "metabase7", "GCF_000175375.1_ASM17537v1_genomic.fasta"), os.path.join(DATADIR, "metabase7", "GCF_000376705.1_ASM37670v1_genomic.fasta"), os.path.join(DATADIR, "metabase7", "GCF_003999335.1_ASM399933v1_genomic.fasta"), os.path.join(DATADIR, "metabase7", "GCF_017638885.1_ASM1763888v1_genomic.fasta"), os.path.join(DATADIR, "metabase7", "GCF_019890915.1_ASM1989091v1_genomic.fasta")]
# MEDIUM20_FASTA = [os.path.join(DATADIR, "metamedium20", "GCF_000005845.2_ASM584v2_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_000009045.1_ASM904v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_000013285.1_ASM1328v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_000025905.1_ASM2590v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_000092125.1_ASM9212v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_000143845.1_ASM14384v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_000159115.1_ASM15911v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_000175375.1_ASM17537v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_000184345.1_ASM18434v2_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_000185725.2_Segn_rugo_CDC_945_V2_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_000196035.1_ASM19603v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_000265425.1_ASM26542v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_000325705.1_ASM32570v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_000376705.1_ASM37670v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_003999335.1_ASM399933v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_008632635.1_ASM863263v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_017638885.1_ASM1763888v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_019704535.1_ASM1970453v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_019890915.1_ASM1989091v1_genomic.fasta"), os.path.join(DATADIR, "metamedium20", "GCF_900638215.1_56553_A01_genomic.fasta")]
# COMPLEX32_FASTA = [os.path.join(DATADIR, 'metacomplex32', 'GCF_000005845.2_ASM584v2_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000006925.2_ASM692v2_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000006945.2_ASM694v2_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000009045.1_ASM904v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000013285.1_ASM1328v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000025905.1_ASM2590v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000092125.1_ASM9212v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000143845.1_ASM14384v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000159115.1_ASM15911v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000160855.1_ASM16085v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000175375.1_ASM17537v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000179955.1_ASM17995v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000184345.1_ASM18434v2_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000185725.2_Segn_rugo_CDC_945_V2_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000196035.1_ASM19603v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000231385.2_ASM23138v3_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000233715.2_ASM23371v3_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000255115.2_ASM25511v3_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000265425.1_ASM26542v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000325705.1_ASM32570v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000376705.1_ASM37670v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_000439255.1_ASM43925v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_001905915.1_ASM190591v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_001932995.2_ASM193299v2_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_003028755.1_ASM302875v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_003999335.1_ASM399933v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_008632635.1_ASM863263v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_017638885.1_ASM1763888v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_018094625.1_ASM1809462v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_019704535.1_ASM1970453v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_019890915.1_ASM1989091v1_genomic.fasta'), os.path.join(DATADIR, 'metacomplex32', 'GCF_900638215.1_56553_A01_genomic.fasta')]
# META_BASE7_ABUNDANCES = os.path.join(META_BASE7_DIR, "nanosim.abundances.tsv")
# MEDIUM20_ABUNDANCES = os.path.join(DATADIR, "meta", "medium20", "nanosim.abundances.tsv")
# COMPLEX32_ABUNDANCES = os.path.join(DATADIR, "meta", "complex32", "nanosim.abundances.tsv")
METAGENOME_ABUNDANCES = os.path.join(DATADIR, "meta", "{metagenome}", "nanosim.abundances.tsv")
METAGENOME_DIR = os.path.join(DATADIR, "meta", "{metagenome}")
# METAGENOME_FASTA = get_metagenome_files("meta{metagenome}", METAGENOME_ABUNDANCES)
METAGENOME_FASTA = os.path.join(DATADIR, "preprocessed_metagenome", "{metagenome}")
REPORT_TEX = os.path.join(REPORTDIR, "output", "{file_name}_k{k}ma{min_abundance}t{threads}", "{report_name}", "{report_file_name}.tex")
REPORT_TEX_FAST = os.path.join(REPORTDIR, "output_fast", "{file_name}_k{k}ma{min_abundance}t{threads}", "{report_name}", "{report_file_name}.tex")
#QUAST_REPORT_TEX = os.path.join(REPORTDIR, "quast_{algorithm}", "{file_name}_k{k}ma{min_abundance}t{threads}", "report.tex")
#QUAST_UNALIGNED = os.path.join(REPORTDIR, "quast_{algorithm}", "{file_name}_k{k}ma{min_abundance}t{threads}", "contigs_reports", "contigs_report_report.unaligned.info")
#QUAST_EXTENDED_REPORT = os.path.join(REPORTDIR, "extended_quast_{algorithm}", "{file_name}_k{k}ma{min_abundance}t{threads}", "report.tex")



#     DATADIR = ... # wherever you have your data, e.g. /wrk-vakka/users/<your username>/flowtigs

#################################
###### Global report rules ######
#################################

# By default, snakemake executes the first rule if no target is given.
# If you don't like that, this rule makes it do nothing if no target is given.
localrules: do_nothing
rule do_nothing:
    shell:  "echo 'No target specified'"

# Here can be rules that specify a set of reports you would like to create, like e.g. all executions for all combinations of different k values and species.
localrules: report_all
rule report_all:
    input:  reports = expand(REPORT, species = ["ecoli", "scerevisiae"], k = [23, 31, 39]),

###############################
###### Report Generation ######
###############################

# Here can be rules to generate reports from your experiments, like e.g. plots or tables.
localrules: create_single_report
rule create_single_report:
    input:  assembly = ASSEMBLY,
    output: report = REPORT,
    conda:  "config/conda-biopython-env.yml"
    script: "scripts/create_single_report.py"


class SafeDict(dict):
    def __missing__(self, key):
        return '{' + key + '}'

def safe_format(str, **kwargs):
    return str.format_map(SafeDict(kwargs))

def safe_expand(str, **kwargs):
    items = []
    for key, values in kwargs.items():
        if type(values) is str or type(values) is not list:
            values = [values]
        items.append([(key, value) for value in values])

    for combination in itertools.product(*items):
        yield safe_format(str, **dict(combination))

def wildcard_format(str, wildcards):
    return str.format(**dict(wildcards.items()))






def get_single_report_script_column_arguments_from_wildcards(wildcards):
    try:
        result = ""
        once = True
        for algorithm in ALGORITHMS:
            if once:
                once = False
            else:
                result += " "

            quast_output_dir = safe_format(QUAST_EXTENDED_OUTPUT_DIR, algorithm = algorithm).format(**wildcards)
            result += f"'{algorithm}' '' '{quast_output_dir}' '' ''"
        return result
    except Exception:
        traceback.print_exc()
        sys.exit("Catched exception")


def get_single_report_script_column_arguments_from_wildcards_fast(wildcards):
    try:
        result = ""
        once = True
        for algorithm in FAST_ALGORITHMS:
            if once:
                once = False
            else:
                result += " "

            quast_output_dir = safe_format(QUAST_EXTENDED_OUTPUT_DIR, algorithm = algorithm).format(**wildcards)
            result += f"'{algorithm}' '' '{quast_output_dir}' '' ''"
        return result
    except Exception:
        traceback.print_exc()
        sys.exit("Catched exception")



rule create_single_report_tex:
    input:  quasts = [safe_format(QUAST_EXTENDED_OUTPUT_DIR, algorithm = algorithm) for algorithm in ALGORITHMS],
            combined_eaxmax_plot = REPORT_COMBINED_EAXMAX_PLOT,
            script = CONVERT_VALIDATION_OUTPUTS_TO_LATEX_SCRIPT,
    output: report = REPORT_TEX,
    log:    log = "logs/create_single_report/{file_name}_k{k}ma{min_abundance}t{threads}r{report_name}rf{report_file_name}/log.log",
    params: genome_name = lambda wildcards: ", ".join(wildcards.file_name), 
            script_column_arguments = get_single_report_script_column_arguments_from_wildcards,
            name_file = REPORT_NAME_FILE,
            hashdir = REPORT_HASHDIR,
    wildcard_constraints:
            report_name = "[^/]+",
    conda: "config/conda-latex-gen-env.yml" 
    threads: 1
    resources:
            queue = "short,medium,bigmem,aurinko",
    shell: """
        mkdir -p '{params.hashdir}'
        echo '{wildcards.report_name} {params.genome_name} {wildcards.report_file_name}' > '{params.name_file}'
        python3 '{input.script}' '{params.hashdir}' '{params.name_file}' 'none' 'none' '{input.combined_eaxmax_plot}' '{output}' {params.script_column_arguments}
        """


rule create_single_report_for_fast_algorithms_only:
    input:  quasts = [safe_format(QUAST_EXTENDED_OUTPUT_DIR, algorithm = algorithm) for algorithm in FAST_ALGORITHMS],
            combined_eaxmax_plot = REPORT_COMBINED_EAXMAX_PLOT_FAST,
            script = CONVERT_FAST_VALIDATION_OUTPUTS_TO_LATEX_SCRIPT,
    output: report = REPORT_TEX_FAST,
    log:    log = "logs/create_single_report/{file_name}_k{k}ma{min_abundance}t{threads}r{report_name}rf{report_file_name}/log.log",
    params: genome_name = lambda wildcards: ", ".join(wildcards.file_name), 
            script_column_arguments = get_single_report_script_column_arguments_from_wildcards_fast,
            name_file = REPORT_NAME_FILE,
            hashdir = REPORT_HASHDIR,
    wildcard_constraints:
            report_name = "[^/]+",
    conda: "config/conda-latex-gen-env.yml" 
    threads: 1
    resources:
            queue = "short,medium,bigmem,aurinko",
    shell: """
        mkdir -p '{params.hashdir}'
        echo '{wildcards.report_name} {params.genome_name} {wildcards.report_file_name}' > '{params.name_file}'
        python3 '{input.script}' '{params.hashdir}' '{params.name_file}' 'none' 'none' '{input.combined_eaxmax_plot}' '{output}' {params.script_column_arguments}
        """


rule create_combined_eaxmax_graph:
    input:  quast_csvs = [os.path.join(safe_format(QUAST_OUTPUT_DIR, algorithm = algorithm), "aligned_stats", "EAxmax_plot.csv") for algorithm in ALGORITHMS],
            script = CREATE_COMBINED_EAXMAX_PLOT_SCRIPT,
    output: REPORT_COMBINED_EAXMAX_PLOT,
    params: input_quast_csvs = lambda wildcards, input: "' '".join([shortname + "' '" + quast for shortname, quast in zip(ALGORITHM_COLUMN_NAMES, input.quast_csvs)])
    conda:  "config/conda-seaborn-env.yml"
    threads: 1
    resources:
            queue = "short,medium,bigmem,aurinko",
    shell: """
        mkdir -p "$(dirname '{output}')"
        python3 '{input.script}' '{params.input_quast_csvs}' '{output}'
        """

rule create_combined_eaxmax_graph_for_fast_algorithms_only:
    input:  quast_csvs = [os.path.join(safe_format(QUAST_OUTPUT_DIR, algorithm = algorithm), "aligned_stats", "EAxmax_plot.csv") for algorithm in FAST_ALGORITHMS],
            script = CREATE_COMBINED_EAXMAX_PLOT_SCRIPT,
    output: REPORT_COMBINED_EAXMAX_PLOT_FAST,
    params: input_quast_csvs = lambda wildcards, input: "' '".join([shortname + "' '" + quast for shortname, quast in zip(FAST_ALGORITHM_COLUMN_NAMES, input.quast_csvs)])
    conda:  "config/conda-seaborn-env.yml"
    threads: 1
    resources:
            queue = "short,medium,bigmem,aurinko",
    shell: """
        mkdir -p "$(dirname '{output}')"
        python3 '{input.script}' '{params.input_quast_csvs}' '{output}'
        """

    

###########################
###### Metagenomes ########
###########################

# Rules to gather the data from metagenomes to a single file


# Rule to get sequences for metagenome data.
localrules: metagenome_to_single_file
rule metagenome_to_single_file:
    input:  abundances = METAGENOME_ABUNDANCES,
            references = METAGENOME_FASTA,
    output: report = METAGENOME,
    log:    log = "logs/metagenome_to_single_file/{metagenome}/log.log",
    conda:  "config/conda-seaborn-env.yml"
    resources:
            time_min = 1440, 
            mem_mb = 10_000, 
            queue = "medium,bigmem,aurinko",
    script: "scripts/metagenome_to_single_file.py"



# Rule to concatenate metagenome data.
localrules: metagenome_concatenate
rule metagenome_concatenate:
    input:  abundances = METAGENOME_ABUNDANCES,
            references = METAGENOME_FASTA,
    output: report = METAGENOME_CONCAT,
    log:    log = "logs/metagenome_concatenate/{metagenome}/log.log",
    conda:  "config/conda-seaborn-env.yml"
    resources:
            time_min = 600, 
            mem_mb = 10_000,
            queue = "medium,bigmem,aurinko",
    script: "scripts/metagenome_concatenate.py"



# # Rule to get sequences for metagenome data.
# localrules: metagenome_to_single_file
# rule metagenome_to_single_file:
#     input:  abundances = META_BASE7_ABUNDANCES,
#             references = META_BASE7_FASTA,
#     output: report = META_BASE7,
#     log:    log = "logs/metagenome_to_single_file/log.log",
#     conda:  "config/conda-seaborn-env.yml"
#     script: "scripts/metagenome_to_single_file.py"



# # Rule to concatenate metagenome data.
# localrules: metagenome_concatenate
# rule metagenome_concatenate:
#     input:  references = META_BASE7_FASTA,
#     output: report = META_BASE7_CONCAT,
#     log:    log = "logs/metagenome_concatenate/log.log",
#     conda:  "config/conda-seaborn-env.yml"
#     script: "scripts/metagenome_concatenate.py"
# HMPJGI73 = [os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Rhodobacter_sphaeroides_2.4.1_plasmid_D,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Deinococcus_radiodurans_R1_plasmid_CP1,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Methanobrevibacter_smithii_ATCC_35061_chromosome,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Rhodobacter_sphaeroides_2.4.1_chromosome_1,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Streptococcus_mutans_UA159_GCA_000007465.2.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Bacillus_cereus_ATCC_10987_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', '_Staphylococcus_aureus_subsp._aureus_USA300__TCH1516_chromosome_.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Bacteroides_vulgatus_ATCC_8482_chromosome,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Rhodobacter_sphaeroides_2.4.1_plasmid_B,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Rhodobacter_sphaeroides_2.4.1_chromosome_2,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Escherichia_coli_str._K-12_substr._MG1655,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Rhodobacter_sphaeroides_2.4.1_plasmid_C,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Rhodobacter_sphaeroides_2.4.1_plasmid_E,_partial_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Clostridium_beijerinckii_NCIMB_8052_chromosome,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Streptococcus_pneumoniae_TIGR4_chromosome,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Rhodobacter_sphaeroides_2.4.1_plasmid_A,_partial_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Candida_albicans_SC5314_GCA_000784635.1.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Actinomyces_odontolyticus_ATCC_17982_Scfld020_genomic_scaffold.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Neisseria_meningitidis_MC58_chromosome,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Helicobacter_pylori_26695_chromosome_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Acinetobacter_baumannii_ATCC_17978_chromosome,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Propionibacterium_acnes_KPA171202_chromosome,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Actinomyces_odontolyticus_ATCC_17982_Scfld021_genomic_scaffold,.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Streptococcus_agalactiae_2603VR_chromosome,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Deinococcus_radiodurans_R1_chromosome_2,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Deinococcus_radiodurans_R1_chromosome_1,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Staphylococcus_epidermidis_ATCC_12228_plasmid_pSE-12228-02,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Acinetobacter_baumannii_ATCC_17978_plasmid_pAB1,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Acinetobacter_baumannii_ATCC_17978_plasmid_pAB2,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Staphylococcus_epidermidis_ATCC_12228_plasmid_pSE-12228-04,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Pseudomonas_aeruginosa_PAO1_chromosome_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Deinococcus_radiodurans_R1_plasmid_MP1,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Enterococcus_faecalis_OG1RF_chromosome,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', '_Staphylococcus_aureus_subsp._aureus_USA300__TCH1516_plasmid_pUSA01-HOU_.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Staphylococcus_epidermidis_ATCC_12228_plasmid_pSE-12228-06,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Bacillus_cereus_ATCC_10987_plasmid_pBc10987,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Staphylococcus_epidermidis_ATCC_12228_plasmid_pSE-12228-03,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Listeria_monocytogenes_EGD-e,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Staphylococcus_epidermidis_ATCC_12228_plasmid_pSE-12228-01,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Lactobacillus_gasseri_ATCC_33323_chromosome,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', '_Staphylococcus_aureus_subsp._aureus_USA300__TCH1516_plasmid_pUSA300HOUMR_.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Staphylococcus_epidermidis_ATCC_12228_plasmid_pSE-12228-05,_complete_sequence.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Staphylococcus_epidermidis_ATCC_12228_chromosome,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'HMP_Mock', 'Streptococcus_mutans_UA159_chromosome,_complete_genome.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Echinicola_vietnamensis_DSM_17526.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Pseudomonas_stutzeri_RCH2.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Meiothermus_silvanus_DSM_9946.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Salmonella_bongori_NCTC_12419.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Natronobacterium_gregoryi_SP2.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Desulfotomaculum_gibsoniae_DSM_7213.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Thermobacillus_composti_KWC4.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Natronococcus_occultus_DSM_3396.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Spirochaeta_smaragdinae_DSM_11293.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Salmonella_enterica_subsp._arizonae_serovar_62_z4_z23_-_strain_RSK2980.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Halovivax_ruber_XH-70.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Olsenella_uli_DSM_7084.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Corynebacterium_glutamicum_ATCC_13032.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Streptococcus_pyogenes_M1_GAS.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Clostridium_thermocellumVPI_7372_ATCC_27405.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Fervidobacterium_pennivorans_DSM_9078.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Coraliomargarita_akajimensis_DSM_45221.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Segniliparus_rotundus_DSM_44985.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Frateuria_aurantia_DSM_6220.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Desulfosporosinus_acidiphilus_SJ4_DSM_22704.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Nocardiopsis_dassonvillei_DSM_43111.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Terriglobus_roseus_DSM_18391.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Hirschia_baltica_ATCC_49814.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'E.coli_K12_ATCC_700926.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Desulfosporosinus_meridiei_DSM_13257.fasta'), os.path.join(DATADIR, "preprocessed_metagenome", 'JGI_Mock', 'Clostridium_perfringensATCC_13124.fasta')]
HMPJGI73 = [os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526015.1_O2.UC52-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527885.1_V1.UC37-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525875.1_O2.UC47-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521185.1_MH0041_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008916905.1_ASM891690v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521275.1_MH0053_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522005.1_MH0125_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525625.1_O2.UC36-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521855.1_MH0110_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523745.1_MH0409_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522635.1_MH0191_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522795.1_MH0220_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526865.1_V1.CD50-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522505.1_MH0177_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521065.1_MH0031_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523935.1_MH0428_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528035.1_V1.UC47-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521785.1_MH0104_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008932165.1_ASM893216v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522775.1_MH0217_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604895.1_ASM360489v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527475.1_V1.UC11-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523255.1_MH0305_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526385.1_V1.CD14-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523225.1_MH0298_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008932195.1_ASM893219v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522895.1_MH0239_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524625.1_MH373_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527405.1_V1.FI34_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524325.1_MH309_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523375.1_MH0326_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521225.1_MH0048_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524615.1_MH372_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525045.1_O2.UC17-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521415.1_MH0067_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525145.1_O2.UC20-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000207925.1_ASM20792v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528275.1_V1.UC59-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525285.1_O2.UC24-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527725.1_V1.UC26-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521445.1_MH0069_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525305.1_O2.UC24-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522135.1_MH0139_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521075.1_MH0032_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523135.1_MH0281_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522555.1_MH0183_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522205.1_MH0144_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528175.1_V1.UC54-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003556005.1_ASM355600v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524765.1_MH394_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604715.1_ASM360471v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527835.1_V1.UC33-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526025.1_O2.UC52-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522575.1_MH0186_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000208165.1_ASM20816v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527645.1_V1.UC21-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527795.1_V1.UC31-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527955.1_V1.UC4-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525225.1_O2.UC22-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524585.1_MH368_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524715.1_MH385_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522985.1_MH0257_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000208065.1_ASM20806v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522325.1_MH0158_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528155.1_V1.UC53-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520885.1_MH0013_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527125.1_V1.FI03_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521365.1_MH0060_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009741655.1_ASM974165v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522935.1_MH0247_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522855.1_MH0234_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526295.1_O2.UC9-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528215.1_V1.UC56-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524545.1_MH361_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523705.1_MH0407_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524845.1_O2.UC11-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527275.1_V1.FI21_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526825.1_V1.CD44-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524975.1_O2.UC14-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521115.1_MH0036_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525805.1_O2.UC43-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521305.1_MH0054_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523575.1_MH0382_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524195.1_MH0455_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525325.1_O2.UC26-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008916805.1_ASM891680v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604615.1_ASM360461v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_011075415.1_ASM1107541v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521315.1_MH0057_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494445.1_ASM2749444v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523565.1_MH0381_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525395.1_O2.UC28-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524655.1_MH377_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525475.1_O2.UC31-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525115.1_O2.UC19-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521105.1_MH0035_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522755.1_MH0211_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521525.1_MH0078_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009685385.1_ASM968538v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000208345.1_ASM20834v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525525.1_O2.UC32-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917245.1_ASM891724v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521725.1_MH0097_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527965.1_V1.UC41-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008916785.1_ASM891678v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_013146465.1_ASM1314646v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494235.1_ASM2749423v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521955.1_MH0120_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520995.1_MH0023_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525575.1_O2.UC35-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525755.1_O2.UC41-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527875.1_V1.UC36-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523835.1_MH0418_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003359075.1_ASM335907v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521345.1_MH0059_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525985.1_O2.UC5-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526045.1_O2.UC53-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523965.1_MH0431_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524775.1_O2.CD1-0-PT_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523625.1_MH0397_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526655.1_V1.CD32-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524985.1_O2.UC15-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526125.1_O2.UC56-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525585.1_O2.UC35-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522885.1_MH0238_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604515.1_ASM360451v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523355.1_MH0321_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526515.1_V1.CD2-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526115.1_O2.UC57-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526765.1_V1.CD41-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003555985.1_ASM355598v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527525.1_V1.UC13-3_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527295.1_V1.FI23_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526575.1_V1.CD25-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523435.1_MH0333_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526465.1_V1.CD18-3_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524405.1_MH338_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520985.1_MH0021_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524835.1_O2.UC1-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527425.1_V1.FI36_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526585.1_V1.CD27-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522785.1_MH0221_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522435.1_MH0169_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521145.1_MH0039_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523555.1_MH0369_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000208145.1_ASM20814v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523665.1_MH0401_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524825.1_O2.CD3-0-PT_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494495.1_ASM2749449v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523675.1_MH0402_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523325.1_MH0316_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524055.1_MH0440_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_011075385.1_ASM1107538v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524225.1_MH0462_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527445.1_V1.UC10-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522925.1_MH0245_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522125.1_MH0138_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522865.1_MH0233_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_013146585.1_ASM1314658v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527035.1_V1.CD-PT4-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523125.1_MH0280_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525705.1_O2.UC40-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523195.1_MH0290_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526285.1_O2.UC9-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525735.1_O2.UC41-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525845.1_O2.UC46-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524345.1_MH311_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524355.1_MH325_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008932185.1_ASM893218v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522465.1_MH0172_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524475.1_MH346_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523515.1_MH0364_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003555605.1_ASM355560v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523715.1_MH0405_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527935.1_V1.UC40-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523975.1_MH0432_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522745.1_MH0206_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523495.1_MH0354_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604875.1_ASM360487v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528075.1_V1.UC49-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524785.1_MH395_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525415.1_O2.UC29-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524685.1_MH383_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521405.1_MH0065_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524185.1_MH0454_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009685395.1_ASM968539v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523015.1_MH0265_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523785.1_MH0413_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520825.1_MH0002_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523885.1_MH0423_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527705.1_V1.UC25-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526425.1_V1.CD17-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604415.1_ASM360441v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003359095.1_ASM335909v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522545.1_MH0181_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522335.1_MH0159_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917045.1_ASM891704v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494315.1_ASM2749431v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524745.1_MH391_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_013146545.1_ASM1314654v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522905.1_MH0243_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521375.1_MH0061_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525975.1_O2.UC5-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520895.1_MH0012_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521595.1_MH0085_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521575.1_MH0074_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521655.1_MH0091_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494455.1_ASM2749445v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523585.1_MH0388_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528095.1_V1.UC5-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522165.1_MH0142_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000208045.1_ASM20804v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_011075405.1_ASM1107540v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521945.1_MH0121_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521265.1_MH0051_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523055.1_MH0268_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526235.1_O2.UC6-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524895.1_O2.UC12-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527085.1_V1.FI02_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528105.1_V1.UC51-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525915.1_O2.UC48-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521485.1_MH0073_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526535.1_V1.CD21-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527805.1_V1.UC31-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008916815.1_ASM891681v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521535.1_MH0079_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524125.1_MH0447_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526335.1_V1.CD1-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526845.1_V1.CD48-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000208025.1_ASM20802v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_013146485.1_ASM1314648v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525135.1_O2.UC19-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524855.1_O2.UC11-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528305.1_V1.UC61-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521135.1_MH0037_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522475.1_MH0174_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527745.1_V1.UC26-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523725.1_MH0406_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494145.1_ASM2749414v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521675.1_MH0092_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527395.1_V1.FI33_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527655.1_V1.UC22-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523315.1_MH0315_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521055.1_MH0030_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528355.1_V1.UC8-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521865.1_MH0111_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521195.1_MH0045_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521965.1_MH0122_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525935.1_O2.UC49-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009741495.1_ASM974149v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522395.1_MH0165_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522085.1_MH0134_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527575.1_V1.UC16-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526145.1_O2.UC58-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522565.1_MH0182_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524235.1_MH0458_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523405.1_MH0330_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521355.1_MH0062_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522025.1_MH0127_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604635.1_ASM360463v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003555845.1_ASM355584v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528055.1_V1.UC48-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526305.1_O2.UC9-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003359055.1_ASM335905v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525105.1_O2.UC19-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_002924505.1_ASM292450v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521425.1_MH0066_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528185.1_V1.UC54-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917085.1_ASM891708v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523035.1_MH0266_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523365.1_MH0322_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524555.1_MH363_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527225.1_V1.FI16_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521495.1_MH0075_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526475.1_V1.CD19-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009741595.1_ASM974159v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524565.1_MH362_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526355.1_V1.CD12-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523175.1_MH0288_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525815.1_O2.UC44-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008916925.1_ASM891692v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526635.1_V1.CD3-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521705.1_MH0095_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526665.1_V1.CD32-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522155.1_MH0141_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526875.1_V1.CD5-0-PN_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523395.1_MH0329_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523645.1_MH0399_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526675.1_V1.CD3-3-PN_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521255.1_MH0052_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526895.1_V1.CD52-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525255.1_O2.UC23-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527585.1_V1.UC17-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525715.1_O2.UC40-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527045.1_V1.CD-PT3-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604525.1_ASM360452v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527025.1_V1.CD-PT1-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917175.1_ASM891717v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003359155.1_ASM335915v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000207965.1_ASM20796v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521585.1_MH0083_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521795.1_MH0103_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527185.1_V1.FI12_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522875.1_MH0236_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494185.1_ASM2749418v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527015.1_V1.CD9-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525055.1_O2.UC17-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523245.1_MH0302_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521715.1_MH0096_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521905.1_MH0115_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604735.1_ASM360473v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523265.1_MH0307_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523095.1_MH0276_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525295.1_O2.UC24-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525365.1_O2.UC27-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494215.1_ASM2749421v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900241105.1_PHH4080.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_013146445.1_ASM1314644v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521325.1_MH0056_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526185.1_O2.UC60-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525015.1_O2.UC15-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524675.1_MH379_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525595.1_O2.UC35-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000208125.1_ASM20812v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527945.1_V1.UC40-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526095.1_O2.UC55-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522605.1_MH0187_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523305.1_MH0314_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524075.1_MH0442_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523655.1_MH0400_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524795.1_O2.CD1-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524305.1_MH293_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521975.1_MH0123_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527715.1_V1.UC25-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525835.1_O2.UC46-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524385.1_MH336_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522385.1_MH0164_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000208325.1_ASM20832v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525745.1_O2.UC41-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522095.1_MH0135_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527625.1_V1.UC19-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526175.1_O2.UC59-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525385.1_O2.UC28-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528005.1_V1.UC45-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522445.1_MH0170_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523085.1_MH0274_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604435.1_ASM360443v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526975.1_V1.CD7-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528285.1_V1.UC60-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521915.1_MH0117_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528065.1_V1.UC49-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521835.1_MH0108_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525535.1_O2.UC32-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527615.1_V1.UC2-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523615.1_MH0396_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527825.1_V1.UC3-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525405.1_O2.UC29-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526735.1_V1.CD38-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525505.1_O2.UC32-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521635.1_MH0088_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524965.1_O2.UC14-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523345.1_MH0319_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522455.1_MH0173_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009741665.1_ASM974166v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523005.1_MH0262_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009741695.1_ASM974169v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526485.1_V1.CD20-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525335.1_O2.UC26-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523795.1_MH0414_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527165.1_V1.FI10_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524455.1_MH345_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525905.1_O2.UC48-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524905.1_O2.UC12-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527155.1_V1.FI09_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524885.1_O2.UC12-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525235.1_O2.UC22-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523895.1_MH0424_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526325.1_V1.CD1-0-PN_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527105.1_V1.FI06_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528015.1_V1.UC4-5_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526815.1_V1.CD45-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521745.1_MH0099_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525545.1_O2.UC3-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520775.1_MH0004_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522915.1_MH0244_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524065.1_MH0441_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522295.1_MH0157_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522145.1_MH0140_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527235.1_V1.FI18_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604535.1_ASM360453v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524035.1_MH0438_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526955.1_V1.CD6-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528245.1_V1.UC58-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522415.1_MH0168_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525685.1_O2.UC38-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526795.1_V1.CD43-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522225.1_MH0147_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523415.1_MH0332_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520835.1_MH0006_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000207945.1_ASM20794v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520815.1_MH0003_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524595.1_MH370_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917165.1_ASM891716v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_013146565.1_ASM1314656v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527375.1_V1.FI31_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527335.1_V1.FI27_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527505.1_V1.UC12-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494335.1_ASM2749433v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526685.1_V1.CD34-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523025.1_MH0264_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520865.1_MH0008_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527345.1_V1.FI28_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009685245.1_ASM968524v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524875.1_O2.UC1-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527315.1_V1.FI24_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524995.1_O2.UC15-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523525.1_MH0366_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000205765.1_ASM20576v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522765.1_MH0212_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000208105.1_ASM20810v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523945.1_MH0430_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525155.1_O2.UC20-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524105.1_MH0445_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524115.1_MH0446_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527845.1_V1.UC34-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524445.1_MH343_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523535.1_MH0365_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524425.1_MH341_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526445.1_V1.CD17-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009741575.1_ASM974157v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523425.1_MH0334_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604855.1_ASM360485v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527905.1_V1.UC38-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521875.1_MH0113_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527205.1_V1.FI14_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522175.1_MH0146_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003369605.1_ASM336960v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525075.1_O2.UC18-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526065.1_O2.UC54-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522285.1_MH0154_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521665.1_MH0090_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527565.1_V1.UC15-3_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521695.1_MH0094_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525865.1_O2.UC46-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527815.1_V1.UC32-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527255.1_V1.FI17_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521985.1_MH0124_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000207985.1_ASM20798v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525825.1_O2.UC44-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524515.1_MH357_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525635.1_O2.UC37-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523735.1_MH0408_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527995.1_V1.UC44-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522065.1_MH0132_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522735.1_MH0205_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524025.1_MH0437_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_002924485.1_ASM292448v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528255.1_V1.UC58-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527555.1_V1.UC15-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527195.1_V1.FI13_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917105.1_ASM891710v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523455.1_MH0347_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525265.1_O2.UC23-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525375.1_O2.UC28-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528325.1_V1.UC63-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525495.1_O2.UC31-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520915.1_MH0015_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604835.1_ASM360483v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523065.1_MH0270_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003555645.1_ASM355564v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524315.1_MH299_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526135.1_O2.UC57-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524735.1_MH390_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526775.1_V1.CD4-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524755.1_MH392_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527465.1_V1.UC1-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522035.1_MH0129_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522115.1_MH0136_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604395.1_ASM360439v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528165.1_V1.UC5-3_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521435.1_MH0068_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526415.1_V1.CD15-3_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009741515.1_ASM974151v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524805.1_O2.CD2-0-PT_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524645.1_MH376_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525695.1_O2.UC40-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526935.1_V1.CD6-0-PN_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526315.1_V1.CD10-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520845.1_MH0001_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917525.1_ASM891752v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526105.1_O2.UC56-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520875.1_MH0011_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525555.1_O2.UC34-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524255.1_MH259_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524575.1_MH367_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524275.1_MH261_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917005.1_ASM891700v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523215.1_MH0297_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521205.1_MH0044_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521085.1_MH0033_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520945.1_MH0018_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526925.1_V1.CD55-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521165.1_MH0043_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522955.1_MH0250_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525195.1_O2.UC21-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525035.1_O2.UC16-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525955.1_O2.UC50-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521825.1_MH0107_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522365.1_MH0163_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604455.1_ASM360445v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520975.1_MH0022_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_013146515.1_ASM1314651v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494345.1_ASM2749434v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523955.1_MH0429_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526605.1_V1.CD29-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522615.1_MH0188_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521045.1_MH0028_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526205.1_O2.UC6-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527535.1_V1.UC14-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522705.1_MH0200_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009685325.1_ASM968532v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525455.1_O2.UC30-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526055.1_O2.UC53-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524935.1_O2.UC13-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527095.1_V1.FI05_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494275.1_ASM2749427v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_013146425.1_ASM1314642v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521245.1_MH0049_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527005.1_V1.CD8-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524285.1_MH271_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527385.1_V1.FI32_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528235.1_V1.UC57-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524155.1_MH0451_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526545.1_V1.CD22-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522715.1_MH0203_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525165.1_O2.UC2-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524495.1_MH355_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524295.1_MH283_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526615.1_V1.CD3-0-PN_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521505.1_MH0077_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522665.1_MH0196_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524465.1_MH350_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527665.1_V1.UC23-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521005.1_MH0025_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522645.1_MH0193_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525085.1_O2.UC18-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522595.1_MH0184_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525965.1_O2.UC50-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521625.1_MH0087_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525895.1_O2.UC47-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521885.1_MH0114_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525565.1_O2.UC34-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523115.1_MH0277_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526435.1_V1.CD16-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526495.1_V1.CD20-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523845.1_MH0419_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003555745.1_ASM355574v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522275.1_MH0153_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521755.1_MH0100_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521545.1_MH0080_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521335.1_MH0058_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527215.1_V1.FI15_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522725.1_MH0204_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527455.1_V1.UC10-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527865.1_V1.UC35-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521475.1_MH0072_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528135.1_V1.UC52-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522845.1_MH0229_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527245.1_V1.FI19_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523545.1_MH0375_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525765.1_O2.UC4-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604375.1_ASM360437v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523765.1_MH0411_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523805.1_MH0415_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526835.1_V1.CD46-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522305.1_MH0156_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526255.1_O2.UC8-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527895.1_V1.UC38-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521555.1_MH0082_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521395.1_MH0063_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524095.1_MH0444_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525355.1_O2.UC27-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526705.1_V1.CD35-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523105.1_MH0275_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525435.1_O2.UC30-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527925.1_V1.UC39-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521895.1_MH0116_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523865.1_MH0421_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527735.1_V1.UC27-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008916945.1_ASM891694v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525205.1_O2.UC2-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523165.1_MH0287_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525065.1_O2.UC17-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522075.1_MH0133_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520965.1_MH0020_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523275.1_MH0308_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604755.1_ASM360475v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525995.1_O2.UC51-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528025.1_V1.UC46-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003556045.1_ASM355604v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526965.1_V1.CD7-0-PN_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526365.1_V1.CD12-3_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523875.1_MH0422_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523485.1_MH0348_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527325.1_V1.FI26_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527365.1_V1.FI30_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526745.1_V1.CD40-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524525.1_MH359_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528315.1_V1.UC62-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528205.1_V1.UC55-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527355.1_V1.FI29_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521215.1_MH0046_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917305.1_ASM891730v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527305.1_V1.FI25_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524535.1_MH360_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526505.1_V1.CD2-0-PN_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520905.1_MH0014_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522315.1_MH0155_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526215.1_O2.UC6-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494395.1_ASM2749439v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523815.1_MH0416_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527775.1_V1.UC30-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524705.1_MH386_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527495.1_V1.UC12-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524485.1_MH349_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000205785.1_ASM20578v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523775.1_MH0412_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003359135.1_ASM335913v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525465.1_O2.UC3-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521465.1_MH0071_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522105.1_MH0137_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525605.1_O2.UC36-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527055.1_V1.CD-PT2-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527485.1_V1.UC11-5_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525795.1_O2.UC43-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525655.1_O2.UC37-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522405.1_MH0166_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522525.1_MH0179_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523915.1_MH0426_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008916845.1_ASM891684v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524415.1_MH339_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523985.1_MH0433_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_013146505.1_ASM1314650v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523075.1_MH0272_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525775.1_O2.UC4-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522235.1_MH0149_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528195.1_V1.UC55-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527115.1_V1.FI04_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524605.1_MH371_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527175.1_V1.FI11_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527135.1_V1.FI07_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523685.1_MH0403_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522185.1_MH0143_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522535.1_MH0180_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527145.1_V1.FI08_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520955.1_MH0019_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522945.1_MH0251_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521175.1_MH0042_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521285.1_MH0050_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521125.1_MH0038_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522375.1_MH0162_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521515.1_MH0076_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522675.1_MH0197_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522815.1_MH0224_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526715.1_V1.CD36-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524865.1_O2.UC11-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522585.1_MH0185_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525675.1_O2.UC38-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521995.1_MH0126_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524375.1_MH328_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526345.1_V1.CD11-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522265.1_MH0152_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523465.1_MH0340_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494415.1_ASM2749441v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524245.1_MH192_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917205.1_ASM891720v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524145.1_MH0450_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523905.1_MH0425_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522995.1_MH0260_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604655.1_ASM360465v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526885.1_V1.CD51-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526915.1_V1.CD54-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009685335.1_ASM968533v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526755.1_V1.CD4-0-PN_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520935.1_MH0017_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604475.1_ASM360447v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525215.1_O2.UC22-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523285.1_MH0310_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524815.1_O2.CD2-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526155.1_O2.UC58-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525445.1_O2.UC30-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_016844825.1_ASM1684482v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523385.1_MH0323_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527765.1_V1.UC29-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524015.1_MH0436_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524725.1_MH387_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917025.1_ASM891702v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526725.1_V1.CD37-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524505.1_MH358_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494295.1_ASM2749429v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526805.1_V1.CD44-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521295.1_MH0055_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525925.1_O2.UC48-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524435.1_MH342_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525315.1_O2.UC26-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526265.1_O2.UC8-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604695.1_ASM360469v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526645.1_V1.CD31-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526525.1_V1.CD21-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008916885.1_ASM891688v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521565.1_MH0081_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524665.1_MH378_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_016844755.1_ASM1684475v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_000208085.1_ASM20808v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524135.1_MH0448_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525425.1_O2.UC29-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525515.1_O2.UC3-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527675.1_V1.UC23-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522965.1_MH0252_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527265.1_V1.FI20_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522355.1_MH0161_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527695.1_V1.UC2-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524005.1_MH0435_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604915.1_ASM360491v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525645.1_O2.UC37-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523295.1_MH0313_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524925.1_O2.UC13-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523045.1_MH0267_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524215.1_MH0457_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009741625.1_ASM974162v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523185.1_MH0289_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526565.1_V1.CD25-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526905.1_V1.CD53-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522695.1_MH0199_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604795.1_ASM360479v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522495.1_MH0176_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525175.1_O2.UC21-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522625.1_MH0189_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524175.1_MH0453_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527855.1_V1.UC35-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528115.1_V1.UC52-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521935.1_MH0119_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526275.1_O2.UC8-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494515.1_ASM2749451v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521815.1_MH0106_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522825.1_MH0227_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526225.1_O2.UC7-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526595.1_V1.CD28-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917125.1_ASM891712v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523635.1_MH0398_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522805.1_MH0222_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527075.1_V1.FI01_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527915.1_V1.UC39-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604575.1_ASM360457v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526195.1_O2.UC60-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527415.1_V1.FI35_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521775.1_MH0102_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525945.1_O2.UC49-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522255.1_MH0151_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526995.1_V1.CD8-0-PN_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521765.1_MH0101_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522195.1_MH0145_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521685.1_MH0093_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526625.1_V1.CD30-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528345.1_V1.UC7-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526785.1_V1.CD42-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521615.1_MH0086_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604815.1_ASM360481v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_013146405.1_ASM1314640v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527545.1_V1.UC14-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521035.1_MH0024_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528365.1_V1.UC9-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527065.1_V1.CD-PT6-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604345.1_ASM360434v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523145.1_MH0284_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521805.1_MH0105_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524695.1_MH384_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522055.1_MH0130_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526375.1_V1.CD13-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522015.1_MH0128_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525245.1_O2.UC2-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917505.1_ASM891750v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009741535.1_ASM974153v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008916985.1_ASM891698v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522485.1_MH0175_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526455.1_V1.CD18-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009685435.1_ASM968543v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526695.1_V1.CD35-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527515.1_V1.UC13-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527595.1_V1.UC17-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526245.1_O2.UC7-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524265.1_MH246_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523155.1_MH0286_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521025.1_MH0026_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525095.1_O2.UC18-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523855.1_MH0420_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524335.1_MH303_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522835.1_MH0230_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521605.1_MH0084_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523205.1_MH0292_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520855.1_MH0007_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008916865.1_ASM891686v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520795.1_MH0005_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524635.1_MH374_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527635.1_V1.UC21-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523505.1_MH0356_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525885.1_O2.UC47-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522215.1_MH0148_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520805.1_MH0009_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520925.1_MH0016_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524915.1_O2.UC1-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527685.1_V1.UC24-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526035.1_O2.UC5-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523755.1_MH0410_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525125.1_O2.UC20-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525855.1_O2.UC44-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521845.1_MH0109_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525785.1_O2.UC43-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528145.1_V1.UC53-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528295.1_V1.UC6-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522425.1_MH0167_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521645.1_MH0089_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526085.1_O2.UC55-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526005.1_O2.UC51-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521455.1_MH0070_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528335.1_V1.UC64-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523335.1_MH0318_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528085.1_V1.UC50-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527605.1_V1.UC18-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917035.1_ASM891703v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604495.1_ASM360449v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523825.1_MH0417_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523995.1_MH0434_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526855.1_V1.CD49-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524205.1_MH0456_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528125.1_V1.UC51-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526165.1_O2.UC59-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523695.1_MH0404_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523445.1_MH0335_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003555805.1_ASM355580v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526395.1_V1.CD1-3-PN_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527985.1_V1.UC43-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604675.1_ASM360467v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527785.1_V1.UC3-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521735.1_MH0098_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522975.1_MH0253_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525025.1_O2.UC16-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525185.1_O2.UC21-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522345.1_MH0160_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525665.1_O2.UC38-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528045.1_V1.UC47-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527755.1_V1.UC28-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526405.1_V1.CD15-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494375.1_ASM2749437v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522655.1_MH0195_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521925.1_MH0118_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525725.1_O2.UC4-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524045.1_MH0439_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522245.1_MH0150_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527975.1_V1.UC42-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525345.1_O2.UC27-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008916965.1_ASM891696v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523475.1_MH0351_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522685.1_MH0198_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009741525.1_ASM974152v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524395.1_MH337_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523925.1_MH0427_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526075.1_O2.UC54-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524165.1_MH0452_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522515.1_MH0178_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_009685425.1_ASM968542v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523235.1_MH0301_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523595.1_MH0393_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526945.1_V1.CD6-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527435.1_V1.FI37_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525615.1_O2.UC36-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525485.1_O2.UC31-1_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900527285.1_V1.FI22_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900525275.1_O2.UC23-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604775.1_ASM360477v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003555905.1_ASM355590v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494245.1_ASM2749424v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524365.1_MH327_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003359115.1_ASM335911v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_003604595.1_ASM360459v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521235.1_MH0047_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524945.1_O2.UC13-2_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900520785.1_MH0010_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_008917135.1_ASM891713v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526555.1_V1.CD24-0_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521385.1_MH0064_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900524085.1_MH0443_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900523605.1_MH0389_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900526985.1_V1.CD7-4_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900521155.1_MH0040_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_027494435.1_ASM2749443v1.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900522045.1_MH0131_assemblies.fasta'), os.path.join(DATADIR, 'preprocessed_metagenome', 'Human_gut', 'GCA_900528265.1_V1.UC59-0_assemblies.fasta')]
# HMPJGI73 = [os.path.join(DATADIR, 'meta73_species', 'Ruminococcus_obeum_ATCC_29174.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_spDOT_SS2--1.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_ovatus_ATCC_8483.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_spDOT_7_2_43FAA.fasta'), os.path.join(DATADIR, 'meta73_species', 'Blautia_hydrogenotrophica_DSM_10507.fasta'), os.path.join(DATADIR, 'meta73_species', 'Eubacterium_dolichum_DSM_3991.fasta'), os.path.join(DATADIR, 'meta73_species', 'Collinsella_stercoris_DSM_13279.fasta'), os.path.join(DATADIR, 'meta73_species', 'Dorea_formicigenerans_ATCC_27755.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_ramosum_DSM_1402.fasta'), os.path.join(DATADIR, 'meta73_species', 'Eubacterium_siraeum_DSM_15702.fasta'), os.path.join(DATADIR, 'meta73_species', 'Coprococcus_comes_ATCC_27758.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_spDOT_4_3_47FAA.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_spDOT_1_1_6.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_cellulosilyticus_DSM_14838.fasta'), os.path.join(DATADIR, 'meta73_species', 'Eubacterium_ventriosum_ATCC_27560.fasta'), os.path.join(DATADIR, 'meta73_species', 'Desulfovibrio_piger_ATCC_29098.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_spDOT_M62--1.fasta'), os.path.join(DATADIR, 'meta73_species', 'Dorea_longicatena_DSM_13814.fasta'), os.path.join(DATADIR, 'meta73_species', 'Ruminococcus_gnavus_ATCC_29149.fasta'), os.path.join(DATADIR, 'meta73_species', 'Parvimonas_micra_ATCC_33270.fasta'), os.path.join(DATADIR, 'meta73_species', 'Roseburia_inulinivorans_DSM_16841.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_uniformis_ATCC_8492.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_coprocola_DSM_17136.fasta'), os.path.join(DATADIR, 'meta73_species', 'Alistipes_putredinis_DSM_17216.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_spiroforme_DSM_1552.fasta'), os.path.join(DATADIR, 'meta73_species', 'Parabacteroides_merdae_ATCC_43184.fasta'), os.path.join(DATADIR, 'meta73_species', 'Collinsella_aerofaciens_ATCC_25986.fasta'), os.path.join(DATADIR, 'meta73_species', 'Coprococcus_eutactus_ATCC_27759.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_spDOT_2_2_4.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_plebeius_DSM_17135.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_spDOT_3_2_5.fasta'), os.path.join(DATADIR, 'meta73_species', 'Blautia_hansenii_DSM_20583.fasta'), os.path.join(DATADIR, 'meta73_species', 'Holdemania_filiformis_DSM_12042.fasta'), os.path.join(DATADIR, 'meta73_species', 'Faecalibacterium_prausnitzii_A2-165.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_caccae_ATCC_43185.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_dorei_5_1_36--D4.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_finegoldii_DSM_17565.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_dorei_DSM_17855.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_spDOT_L2-50.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_spDOT_9_1_42FAA.fasta'), os.path.join(DATADIR, 'meta73_species', 'Parabacteroides_johnsonii_DSM_18315.fasta'), os.path.join(DATADIR, 'meta73_species', 'Eubacterium_biforme_DSM_3989.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_hylemonae_DSM_15053.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_bartlettii_DSM_16795.fasta'), os.path.join(DATADIR, 'meta73_species', 'Ruminococcus_torques_ATCC_27756.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_spDOT_D1.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_difficile_NAP08.fasta'), os.path.join(DATADIR, 'meta73_species', 'Anaerotruncus_colihominis_DSM_17241.fasta'), os.path.join(DATADIR, 'meta73_species', 'Eubacterium_hallii_DSM_3353.fasta'), os.path.join(DATADIR, 'meta73_species', 'Prevotella_copri_DSM_18205.fasta'), os.path.join(DATADIR, 'meta73_species', 'Anaerostipes_caccae_DSM_14662.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_intestinalis_DSM_17393.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_stercoris_ATCC_43183.fasta'), os.path.join(DATADIR, 'meta73_species', 'Faecalibacterium_prausnitzii_M21--2.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_bolteae_ATCC_BAA-613.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_nexile_DSM_1787.fasta'), os.path.join(DATADIR, 'meta73_species', 'Catenibacterium_mitsuokai_DSM_15897.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_spDOT_D2.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_fragilis_3_1_12.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_scindens_ATCC_35704.fasta'), os.path.join(DATADIR, 'meta73_species', 'Butyrivibrio_crossotus_DSM_2876.fasta'), os.path.join(DATADIR, 'meta73_species', 'Ruminococcus_lactaris_ATCC_29176.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_coprophilus_DSM_18228.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_methylpentosum_DSM_5476.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_hiranonis_DSM_13275.fasta'), os.path.join(DATADIR, 'meta73_species', 'Bacteroides_eggerthii_DSM_20697.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_hathewayi_DSM_13479.fasta'), os.path.join(DATADIR, 'meta73_species', 'Mitsuokella_multacida_DSM_20544.fasta'), os.path.join(DATADIR, 'meta73_species', 'Collinsella_intestinalis_DSM_13280.fasta'), os.path.join(DATADIR, 'meta73_species', 'Roseburia_intestinalis_L1-82.fasta'), os.path.join(DATADIR, 'meta73_species', 'Subdoligranulum_variabile_DSM_15176.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_asparagiforme_DSM_15981.fasta'), os.path.join(DATADIR, 'meta73_species', 'Clostridium_leptum_DSM_753.fasta')]

rule many_rules:
    input: HMPJGI73
    output: [os.path.join(DATADIR, "test")]
    shell: """
    ls
    """


########################
###### Assembly ########
########################


# Simulates abundances for a metagenome using the Poisson distribution
rule generate_abundances:
    input: reads = METAGENOME_DIR,
    output: abundances = METAGENOME_ABUNDANCES,
    log:    log = "logs/preprocessing_single_genome/meta_{metagenome}/log.log",
    conda:  "config/conda-biopython-env.yml"
    script: "scripts/calculate_abundances.py"


# Rule to remove all the non-[A, C, G, T] characters from a fasta file. Should be ran first.
rule preprocessing_single_genome:
    input:  assembly = UNPREPROCESSED_METAGENOME,
    output: report = PREPROCESSED_METAGENOME,
    log:    log = "logs/preprocessing_single_genome/{metagenome}/{genome}/log.log",
    conda:  "config/conda-biopython-env.yml"
    resources:
            mem_mb = 50_000, 
            time_min = 120,
            cpus = build_cpus,
            queue = 'aurinko,bigmem,short,medium', 
    
    script: "scripts/preprocessing.py"


# Rule to remove all the non-[A, C, G, T] characters from a fasta file. Should be ran first.
rule preprocessing_all_genomes:
    input:  assembly = METAGENOME_DIR,
    output: report = directory(METAGENOME_FASTA),
    log:    log = "logs/preprocessing_single_genome/{metagenome}/log.log",
    conda:  "config/conda-biopython-env.yml"
    resources:
            time_min = 1440, 
            mem_mb = 10_000, # likely too much
            queue = "medium,bigmem,aurinko",
    script: "scripts/preprocessing_all_genomes.py"


# Rule to circularize non-circular sequences of the genome or metagenome.
# input: data of the non-circular sequences of the genome or metagenome in fasta format.
# wildcards: k=size of kmers, file_name=name of the file with the data in the data folder,
#   min_abundance=minimum abundance, threds=number of cpu cores used.
# output: data of the circular sequences of the genome or metagenome in fasta format.
localrules: circularization
rule circularization:
    input:  assembly = GENOME_ALL_REFERENCES,
    output: report = GENOME_CIRCULAR_REFERENCES,
    log:    log = "logs/circularization/{file_name}_k{k}ma{min_abundance}t{threads}/log.log",
    conda:  "config/conda-biopython-env.yml"
    script: "scripts/circularization.py"
399308

# Rule to create a node node-cenric weighted De Bruijn graph using the bcalm2 tool.
# input: data of the circular sequences of the genome or metagenome in fasta format.
# wildcards: k=size of kmers, file_name=name of the file with the data in the data folder,
#   min_abundance=minimum abundance, threds=number of cpu cores used.
# output: Node-centric weighted De Bruijn graph with k-mers of size k in fasta format.
rule bcalm2_build:
    input:  references = GENOME_CIRCULAR_REFERENCES,
    output: tigs = BUILD_FA,
    log:    log = "logs/bcalm2/{file_name}_k{k}ma{min_abundance}t{threads}/log.log",
    params: references = lambda wildcards, input: "'" + "' '".join(input.references) + "'",
    conda:  "config/conda-bcalm2-env.yml",
    threads: MAX_THREADS,
    shadow: "minimal"
    resources:
            mem_mb = 10_000, # probably much more than needed
            time_min = 60,
            cpus = build_cpus,
            queue = 'aurinko,bigmem,short,medium', # I had some more complex expression here, the queues fitting to the time are on https://wiki.helsinki.fi/display/it4sci/HPC+Environment+User+Guide#HPCEnvironmentUserGuide-4.8.4Partitions-Ukko
    shell:  """
        rm -f '{log.log}'
        ${{CONDA_PREFIX}}/bin/time -v bcalm -nb-cores {threads} -kmer-size {wildcards.k} -in '{input.references}' -out '{output.tigs}' -abundance-min {wildcards.min_abundance} 2>&1 | tee -a '{log.log}'
        mv '{output.tigs}.unitigs.fa' '{output.tigs}'
        """


# Rule to set up the external software for the node_to_arc_centric_dbg rule.
rule build_node_to_arc_centric_dbg:
    input:  "external_software/node-to-arc-centric-dbg/Cargo.toml",
    output: NODE_TO_ARC_CENTRIC_DBG_BINARY,
    conda:  "config/conda-rust-env.yml",
    threads: MAX_THREADS,
    resources:
            mem_mb = 10000,
            time_min = 60,
            cpus = MAX_THREADS,
            queue = "aurinko,bigmem,short,medium",
    shell:  """
        cd external_software/node-to-arc-centric-dbg
        cargo build --release -j {threads} --offline
    """


# Rule to transform the node-centric weighted De Bruijn graph given by the bcalm2_build
#   rule to an arc-centric weighted De Bruijn graph.
# input: output of bcalm2_build and output of build_node_to_arc_centric_dbg.
# wildcards: k=size of kmers, file_name=name of the file with the data in the data folder,
#   min_abundance=minimum abundance, threads=number of cpu cores used.
# output: Arc-centric weighted De Bruijn graph with k-mers of size k in fasta format.
rule node_to_arc_centric_dbg:
    input:  node_centric_dbg = BUILD_FA,
            binary = NODE_TO_ARC_CENTRIC_DBG_BINARY,
    log:    log = "logs/node_to_arc/{file_name}_k{k}ma{min_abundance}t{threads}/log.log",
    output: arc_centric_dbg = NODE_TO_ARC_CENTRIC_DBG,   # .edgelist
    conda:  "config/conda-time-env.yml",
    resources:
            time_min = 60, # likely too much
            mem_mb = 10_000, # likely too much
            queue = "short,medium,bigmem,aurinko",
    shell:  """
        rm -f '{log.log}'
        ${{CONDA_PREFIX}}/bin/time -v '{input.binary}' -k {wildcards.k} --input '{input.node_centric_dbg}' --output '{output.arc_centric_dbg}'
    """




# Rule to set up the external software for the safe-paths rule.
rule build_safe_paths:
    input:  "external_software/safe-paths/Cargo.toml",
    output: SAFE_PATHS_BINARY,
    conda:  "config/conda-rust-env.yml",
    threads: MAX_THREADS,
    resources:
            mem_mb = 10000,
            time_min = 60,
            cpus = MAX_THREADS,
            queue = "aurinko,bigmem,short,medium",
    shell:  """
        cd external_software/safe-paths
        cargo build --release -j {threads} --offline
    """


# Rule to get the safe paths from the edge-centric weighted De Bruijn graph given by
#   the node_to_arc_centric_dbg rule.
# input: output of node_to_arc_centric_dbg and output of build_safe_paths.
# wildcards: k=size of kmers, file_name=name of the file with the data in the data folder,
#   min_abundance=minimum abundance, thredas=number of cpu cores used.
# output: List of the safe paths with, for each path, a line with ">Path_<path-index>"
#   and a second line with the sequence.
rule safe_paths:
    input:  arc_centric_dbg = NODE_TO_ARC_CENTRIC_DBG,
            binary = SAFE_PATHS_BINARY,
    log:    log = "logs/safe_paths/{file_name}_k{k}ma{min_abundance}t{threads}/log.log",
    output: safe_paths = SAFE_PATHS,
    conda:  "config/conda-time-env.yml",
    resources:
            time_min = 60, # likely too much
            mem_mb = 100_000, # likely too much
            queue = "medium,bigmem,aurinko",
    shell:  """
        rm -f '{log.log}'
        ${{CONDA_PREFIX}}/bin/time -v '{input.binary}' -k {wildcards.k} --input '{input.arc_centric_dbg}' --output '{output.safe_paths}'
    """




# Rule to make a report of the results.
# input: 
#   contigs: List of safe paths in fasta format.
#   script: File to execute the Rust program.
#   error_report: Template file for when there is an error creating the report.tex file.
#   error_missassemblies_report: Template file for when there is an error creating the missassemblies_report.tex file.
# wildcards: 
#   algorithm: method for finding safe paths. Either "uni" for unitigs, "omni" for omnitigs or "flow" for flowtigs.
#   k: size of kmers
#   file_name: name of the file with the data in the data folder
#   min_abundance: minimum abundance
#   threads: number of cpu cores used.
# output: report in .tex, .tsv, .txt, and .pdf formats
rule run_quast:
    input:  contigs = os.path.join(REPORTDIR, "safe_paths_{algorithm}", "{file_name}_k{k}ma{min_abundance}t{threads}", "report.fasta"),
            references = [GENOME_CONCAT_REFERENCES], # list of references
            script = QUAST_BINARY,
            error_report = "config/quast_error_report.tex",
            error_misassemblies_report = "config/quast_error_misassemblies_report.tex"
    output: directory = directory(QUAST_OUTPUT_DIR),
            eaxmax_csv = os.path.join(QUAST_OUTPUT_DIR, "aligned_stats/EAxmax_plot.csv"),
    log:    minimap = os.path.join(QUAST_OUTPUT_DIR, "contigs_reports", "contigs_report_contigs.stderr")
    params: references = lambda wildcards, input: "-r '" + "' -r '".join(input.references) + "'",
    conda: "config/conda-quast-env.yml"
    threads: 14,
    resources: mem_mb = 1_000_000, # likely to much for our genomes
               cpus = 14,
               time_min = 1440,
               queue = "bigmem,aurinko", # I had some more complex expression here, the queues fitting to the time are on https://wiki.helsinki.fi/display/it4sci/HPC+Environment+User+Guide#HPCEnvironmentUserGuide-4.8.4Partitions-Ukko
    shell:  """
        set +e 
        ${{CONDA_PREFIX}}/bin/time -v {input.script} -t {threads} --no-html -o '{output.directory}' {params.references} '{input.contigs}'
        set -e
        
        if [ $? -ne 0 ]; then
            rm -rf '{output.directory}'
        fi


        mkdir -p '{output.directory}'
        if [ ! -f '{output.directory}/report.tex' ]; then
            echo "report.tex is missing, using error template" 
            cp '{input.error_report}' '{output.directory}/report.tex'
        fi

        mkdir -p '{output.directory}/contigs_reports'
        if [ ! -f '{output.directory}/contigs_reports/misassemblies_report.tex' ]; then
            echo "misassemblies_report.tex is missing, using error template"
            cp '{input.error_misassemblies_report}' '{output.directory}/contigs_reports/misassemblies_report.tex'
        fi

        mkdir -p '{output.directory}/aligned_stats'
        if [ ! -f '{output.directory}/aligned_stats/EAxmax_plot.csv' ]; then
            echo "EAxmax_plot.csv is missing, using empty file"
            touch '{output.directory}/aligned_stats/EAxmax_plot.csv'
        fi
    """ 


# Rule to add the length of the longest unaligned contig to the quast report
rule add_longest_unaligned_contig_length_to_report:
    input: directory = QUAST_OUTPUT_DIR,
           script = "scripts/longest_unaligned_contig.py",
    output: directory(QUAST_EXTENDED_OUTPUT_DIR)
    log:    log = "logs/the_longest_unaligned_{algorithm}/{file_name}_k{k}ma{min_abundance}t{threads}/log.log",
    conda:  "config/conda-seaborn-env.yml"
    shell:"""
        mkdir -p {output}
        cp -r {input.directory}/* {output}
        mv '{output}/report.tex' '{output}/old_report.tex'
        touch '{output}/report.tex'
        ${{CONDA_PREFIX}}/bin/time -v '{input.script}' '{output}/old_report.tex' '{output}/contigs_reports/contigs_report_report.unaligned.info' '{output}/report.tex'
    """




# Rule to convert a tex file to a pdf
rule latex_to_pdf:
    input: "{subpath}{file_suffix}.tex"
    output: "{subpath}{file_suffix}.pdf"
    wildcard_constraints:
        file_suffix = "((report)|(all)|(reduced))",
    conda: "config/conda-latex-env.yml"
    threads: 1
    resources:
            queue = "short,medium,bigmem,aurinko",
    shell: """
        tectonic '{input}'
        """




# Rule to set up the external software for the practical omnitigs rule.
rule build_practical_omnitigs:
    input:  "external_software/practical-omnitigs/implementation/Cargo.toml",
    output: PRACTICAL_OMNITIGS_BINARY,
    conda:  "config/conda-practical-omnitigs-env.yml",
    threads: MAX_THREADS,
    resources:
            mem_mb = 10000,
            time_min = 60,
            cpus = MAX_THREADS,
            queue = "aurinko,bigmem,short,medium",
    shell:  """
        cd external_software/practical-omnitigs/implementation
        cargo build --release -j {threads} 
    """



# Rule to get the safe paths from the node-centric weighted De Bruijn graph given by
#   the bcalm2 rule.
# input: output of bcalm2 and output of build_practical_omnitigs.
# wildcards: k=size of kmers, file_name=name of the file with the data in the data folder,
#   min_abundance=minimum abundance, thredas=number of cpu cores used.
# output: List of the safe paths with, for each path, a line with "><path-index>"
#   and a second line with the sequence.
rule practical_multisafe:
    input:  practical_omnitigs = BUILD_FA,
            binary = PRACTICAL_OMNITIGS_BINARY,
    log:    log = "logs/practical_omnitigs/{file_name}_k{k}ma{min_abundance}t{threads}/log.log",
    output: practical_omnitigs = PRACTICAL_OMNITIGS,  
    conda:  "config/conda-rust-env.yml",
    resources:
            time_min = 4320, 
            mem_mb = 100_000, # likely too much
            queue = "bigmem,aurinko",
    shell:  """
        rm -f '{log.log}'
        '{input.binary}' compute-multi-safe-walks --file-format bcalm2 --input '{input.practical_omnitigs}' --output '{output.practical_omnitigs}' -k {wildcards.k}
    """


# Rule for testing the practical omnitigs rule
rule practical_omitigs:
    input:  practical_omnitigs = BUILD_FA,
            binary = PRACTICAL_OMNITIGS_BINARY,
    log:    log = "logs/practical_test_omnitigs/{file_name}_k{k}ma{min_abundance}t{threads}/log.log",
    output: practical_omnitigs = PRACTICAL_TEST_OMNITIGS,  
    conda:  "config/conda-rust-env.yml",
    resources:
            time_min = 4320, 
            mem_mb = 100_000, # likely too much
            queue = "bigmem,aurinko",
    shell:  """
        rm -f '{log.log}'
        '{input.binary}' compute-omnitigs --file-format bcalm2 --input '{input.practical_omnitigs}' --output '{output.practical_omnitigs}' -k {wildcards.k}
    """


# Rule for testing the practical omnitigs rule
rule practical_trivial_omitigs:
    input:  practical_omnitigs = BUILD_FA,
            binary = PRACTICAL_OMNITIGS_BINARY,
    log:    log = "logs/practical_trivial_omnitigs/{file_name}_k{k}ma{min_abundance}t{threads}/log.log",
    output: practical_omnitigs = PRACTICAL_TRIVIAL_OMNITIGS,  
    conda:  "config/conda-rust-env.yml",
    resources:
            time_min = 60, # likely too much
            mem_mb = 100_000, # likely too much
            queue = "medium,bigmem,aurinko",
    shell:  """
        rm -f '{log.log}'
        '{input.binary}' compute-trivial-omnitigs --non-scc --multi-safe --file-format bcalm2 --input '{input.practical_omnitigs}' --output '{output.practical_omnitigs}' -k {wildcards.k}
    """


# Rule for running multiple pipelines at the same time. Insert the outputs of the pipelines that you want to run in the input of this rule.
rule run_multiple_pipelines:
    input: pipeline_outputs = [os.path.join(REPORTDIR, "output", "meta_HMP_Mock_k31ma1t28", "report_thu17-8", "report.tex"), os.path.join(REPORTDIR, "output", "meta_JGI_Mock_k31ma1t28", "report_thu17-8", "report.tex"), os.path.join(REPORTDIR, "output", "meta_73_species_k31ma1t28", "report_thu17-8", "report.tex")]
    output: empty_file = os.path.join(REPORTDIR, "multiple_runs")
    shell:  """
        touch multiple_runs
    """


#######################
###### Downloads ######
#######################

# Here can be rules to download your input files.
# These must be localrules, as the compute nodes of the federated cluster do not have a connection to the internet.

def get_reads_url(wildcards):
    # Snakemake just swallows whatever errors happen in functions.
    # The only way around that is to handle and print the error ourselves.
    try:
        species = wildcards.species

        if species == "ecoli":
            return "https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos3/sra-pub-run-20/SRR12793243/SRR12793243.1"
        elif species == "scerevisiae":
            return "https://sra-download.ncbi.nlm.nih.gov/traces/sra78/SRR/014398/SRR14744387"
        else:
            raise Exception("Unknown species: {}".format(species))
    except Exception:
        traceback.print_exc()
        sys.exit("Catched exception")


# Rule to download the E.coli bacteria data
localrules: download_ecoli
rule download_ecoli:
    output: ECOLI
   # conda:  "config/conda-rust-env.yml"
    threads: 1
    shell:  """
        mkdir -p data
        cd data

        rm -rf ecoli.fasta
        wget -O ecoli.fasta.gz https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/005/845/GCF_000005845.2_ASM584v2/GCF_000005845.2_ASM584v2_genomic.fna.gz
        gunzip ecoli.fasta.gz
    """

# Rule needed for run_quast. Just copies the file with a new name.
rule ecoli_concat:
    input: file = ECOLI
    output: file = ECOLI_CONCAT
    shell: """
        cp {input.file} {output.file}
    """

# Rule to create the the concatenated version of a single file. Just copies the file with a new name.
rule single_concat:
    input: file = SINGLE
    output: file = SINGLE_CONCAT
    shell: """
        cp {input.file} {output.file}
    """


# Rule to download the external software used for turning the node-centric De Bruijn graph
#   to an arc-centric one.
localrules: download_node_to_arc_centric_dbg
rule download_node_to_arc_centric_dbg:
    output: "external_software/node-to-arc-centric-dbg/Cargo.toml"
    conda:  "config/conda-rust-env.yml"
    threads: 1
    shell:  """
        mkdir -p external_software
        cd external_software

        rm -rf node-to-arc-centric-dbg
        git clone https://github.com/sebschmi/node-to-arc-centric-dbg.git
        cd node-to-arc-centric-dbg
        git checkout 9ae1e61bd089be227342d03762031381cb989888 
        cargo fetch
    """ 


# Rule to download the external software used for calculatings safe paths from an arc-centric De Bruijn graph.
localrules: download_safe_paths
rule download_safe_paths:
    output: "external_software/safe-paths/Cargo.toml"
    conda:  "config/conda-rust-env.yml"
    threads: 1
    shell:  """
        mkdir -p external_software
        cd external_software

        rm -rf safe-paths
        git clone https://github.com/elieling/safe-paths.git
        cd safe-paths
        git checkout 19304e99283e336ebf15d5b206203da00b053f82

        cargo fetch
    """ 


localrules: install_quast
rule install_quast:
    output: script = QUAST_BINARY,
    params: external_software_dir = "external_software", 
    threads: 1
    shell: """
        mkdir -p {params.external_software_dir}
        cd {params.external_software_dir}

        rm -rf quast
        git clone https://github.com/elieling/quast.git 
        cd quast
        git checkout d370ad11d60122d11220fc212c44be1be88daf9c
    """ # af809a144d007f8a31cfffee994b33362a3e379e



localrules: download_practical_omnitigs
rule download_practical_omnitigs:
    output: "external_software/practical-omnitigs/implementation/Cargo.toml"
    conda:  "config/conda-practical-omnitigs-env.yml"
    threads: 1
    shell:  """
        mkdir -p external_software
        cd external_software

        rm -rf practical-omnitigs
        git clone https://github.com/algbio/practical-omnitigs.git
        cd practical-omnitigs
        git checkout 16a17703e6cd060863857a2d348cdecb2df12aa5
        cd implementation
        cargo fetch
    """ # f1c451ee2ce59f6e63ca32c77ac0cd590a8d5ecb 






localrules: download_sra_file
rule download_sra_file:
    output: file = READS_SRA,
    params: url = get_reads_url,
    shell:  """
        wget --progress=dot:mega -O '{output.file}' '{params.url}'
    """

# If no resources are specified, the rule will use the default resources specified in config/turso/config.yml
rule convert_sra_download:
    input:  file = READS_SRA,
    output: file = READS_FASTA,
    conda:  "config/conda-convert-reads-env.yml"
    shell:  "fastq-dump --stdout --fasta default '{input.file}' > '{output.file}'"

# I link the reads to each assembly directory separately, such that bcalm creates its auxiliary files in that directory instead of the download directory.
localrules: download_reads
rule download_reads:
    input:  file = READS_FASTA,
    output: file = ASSEMBLY_SOURCE_READS,
    shell:  "ln -sr -T '{input.file}' '{output.file}'"

##############################
###### Download results ######
##############################

# Here is a template for a rule to download report files from turso to your local machine.
# You can use something similar, just make sure to update the paths and the include specification.
#localrules: sync_turso_results
#rule sync_turso_results:
#    conda: "config/conda-rsync-env.yml"
#    shell: """
#        mkdir -p data/reports
#        rsync --verbose --recursive --no-relative --include="*/" --include="report.pdf" --include="aggregated-report.pdf" --exclude="*" turso:'/proj/sebschmi/git/practical-omnitigs/data/reports/' data/reports
#        """

